File: teaching-examples/pinocchio_gravity_comp.py
#!/usr/bin/env python3
"""
Standalone teaching example — NOT part of open_manipulator ROS packages.
Pinocchio gravity compensation g(q) for OpenMANIPULATOR-X arm.

Prerequisites:
  sudo apt install ros-${ROS_DISTRO}-pinocchio   # or pip install pin
  export URDF_PATH=/path/to/open_manipulator_x.urdf

Usage:
  python3 teaching-examples/pinocchio_gravity_comp.py
  python3 teaching-examples/pinocchio_gravity_comp.py /path/to/model.urdf 0 -1 0.7 0.3
"""

import sys
import os

import pinocchio as pin
import numpy as np


def main():
    urdf = os.environ.get(
        "URDF_PATH",
        os.path.join(
            os.path.dirname(__file__),
            "..",
            "open_manipulator_description",
            "urdf",
            "open_manipulator_x",
            "open_manipulator_x.urdf",
        ),
    )
    urdf = os.path.abspath(urdf)

    if len(sys.argv) > 1:
        urdf = sys.argv[1]

    if not os.path.isfile(urdf):
        print(f"URDF not found: {urdf}")
        print("Set URDF_PATH or pass path as argv[1]")
        sys.exit(1)

    # Fixed base: do not add a free-flyer root joint
    model = pin.buildModelFromUrdf(urdf)
    data = model.createData()

    print(f"Model: nq={model.nq}, nv={model.nv}")
    print(f"Joint names: {[model.names[i] for i in range(1, model.njoints)]}")

    # Default q: SRDF home pose for arm, or from CLI
    if len(sys.argv) >= 6:
        q_arm = np.array([float(sys.argv[i]) for i in range(2, 6)])
    else:
        q_arm = np.array([0.0, -1.0, 0.7, 0.3])  # home

    q = pin.neutral(model)
    # Map first 4 velocity DoF (arm) — adjust if your URDF model layout differs
    nv_arm = min(4, model.nv)
    q[:nv_arm] = q_arm[:nv_arm]

    # Gravity vector (default is (0,0,-9.81) in pinocchio)
    print(f"model.gravity.linear = {model.gravity.linear.T}")

    g = pin.computeGeneralizedGravity(model, data, q)
    tau_rnea = pin.rnea(
        model, data, q, np.zeros(model.nv), np.zeros(model.nv)
    )

    print(f"\nq (first {nv_arm} coords) = {q[:nv_arm].T}")
    print(f"g(q) = {g[:nv_arm].T}")
    print(f"rnea(q,0,0) = {tau_rnea[:nv_arm].T}")
    print(f"max |g - rnea| = {np.max(np.abs(g - tau_rnea)):.2e}")

    print("\n→ Command +g(q) on effort interface to compensate gravity (sign may need")
    print("  calibration vs Dynamixel / KDL convention on real hardware).")


if __name__ == "__main__":
    main()

File: teaching-examples/pinocchio_gravity_comp.cpp
/**
 * Standalone teaching example — NOT part of open_manipulator ROS packages.
 * Pinocchio gravity compensation g(q) for a fixed-base URDF manipulator.
 *
 * Build (ROS 2 workspace with pinocchio installed):
 *   g++ -std=c++17 teaching-examples/pinocchio_gravity_comp.cpp -o /tmp/pinocchio_gc \
 *       $(pkg-config --cflags --libs pinocchio)
 *
 * Run:
 *   /tmp/pinocchio_gc path/to/open_manipulator_x.urdf
 */

#include <iostream>
#include <iomanip>

#include <pinocchio/parsers/urdf.hpp>
#include <pinocchio/algorithm/rnea.hpp>
#include <pinocchio/algorithm/joint-configuration.hpp>

int main(int argc, char ** argv)
{
  const std::string urdf_path =
    (argc > 1) ? argv[1] :
    "../open_manipulator_description/urdf/open_manipulator_x/open_manipulator_x.urdf";

  pinocchio::Model model;
  pinocchio::urdf::buildModel(urdf_path, model);
  pinocchio::Data data(model);

  std::cout << "nq=" << model.nq << " nv=" << model.nv << "\n";
  std::cout << "gravity: " << model.gravity.linear().transpose() << "\n";

  Eigen::VectorXd q = pinocchio::neutral(model);
  // SRDF home: joint1..4 = 0, -1, 0.7, 0.3
  if (model.nv >= 4) {
    q[0] = 0.0;
    q[1] = -1.0;
    q[2] = 0.7;
    q[3] = 0.3;
  }

  const Eigen::VectorXd v = Eigen::VectorXd::Zero(model.nv);
  const Eigen::VectorXd a = Eigen::VectorXd::Zero(model.nv);

  pinocchio::computeGeneralizedGravity(model, data, q);
  const Eigen::VectorXd & g = data.g;

  pinocchio::rnea(model, data, q, v, a);
  const Eigen::VectorXd & tau = data.tau;

  std::cout << std::fixed << std::setprecision(4);
  std::cout << "q      = " << q.head(std::min<int>(4, model.nq)).transpose() << "\n";
  std::cout << "g(q)   = " << g.head(std::min<int>(4, model.nv)).transpose() << "\n";
  std::cout << "rnea   = " << tau.head(std::min<int>(4, model.nv)).transpose() << "\n";
  std::cout << "||g-rnea||_inf = " << (g - tau).lpNorm<Eigen::Infinity>() << "\n";

  return 0;
}
